In [5]:
import pandas as pd
df_sales = pd.read_csv('csv/sales.csv')
print(df_sales.shape)
print(df_sales.head())
df_customers = pd.read_csv('csv/customers.csv')
print(df_customers.shape)
print(df_customers.head())
df = pd.merge(df_sales, df_customers, on='customer_id', how='left')
print(df.shape)
print(df.dtypes)
print(df.head())

(15234, 11)
   sale_id        invoice_id   sale_date  customer_id  product_id  quantity  \
0        1  INV-202301-00001  2023-01-10       2588.0        1274         2   
1        2  INV-202301-00002  2023-01-16       4338.0        1207         2   
2        3  INV-202301-00003  2023-01-05       4673.0        1264         1   
3        4  INV-202301-00004  2023-01-02       4677.0        1341         3   
4        5  INV-202301-00005  2023-01-13       2390.0        1284         1   

   unit_price  total_price channel store_location payment_method  
0      234.79       469.58    pood        Tallinn          kaart  
1      241.13       482.26    pood          Pärnu      järelmaks  
2      258.46       221.19    pood          Pärnu      järelmaks  
3       45.21       135.63    pood          Tartu       sularaha  
4       99.57        99.57    pood          Tartu          kaart  
(3150, 9)
   customer_id first_name last_name                   email           phone  \
0         2001        

In [ ]:
df.isna().sum()

In [5]:
df.isna().sum()

sale_id                 0
invoice_id              0
sale_date               0
customer_id          1487
product_id              0
quantity                0
unit_price              0
total_price             0
channel                 0
store_location       5204
payment_method          0
first_name           1487
last_name            1487
email                2905
phone                1487
city                 1487
registration_date    1487
loyalty_tier         7033
birth_year           1487
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(4086)

In [7]:
df[df.duplicated()].head()

,sale_id,invoice_id,sale_date,customer_id,product_id,quantity,unit_price,total_price,channel,store_location,payment_method,first_name,last_name,email,phone,city,registration_date,loyalty_tier,birth_year
10118,6957,INV-202407-00266,2024-07-05,4909.0,1084,1,156.13,156.13,pood,Tallinn,järelmaks,Mihkel,Talvik,mihkel.talvik2@hot.ee,+372 5128 9327,Tartu,2021-05-12,silver,1996.0
10120,506,INV-202302-00245,2023-02-16,4307.0,1315,1,95.50,95.50,pood,Pärnu,kaart,Merike,Rebane,merike.rebane@outlook.com,+372 5126 9930,Kuressaare,2022-07-01,NaN,1998.0
10121,3791,INV-202311-00286,2023-11-29,4701.0,1343,3,63.87,191.61,pood,Tallinn,järelmaks,Enn,Rebane,NaN,+372 5952 5043,Tartu,2022-12-16,NaN,1967.0
10122,6982,INV-202407-00291,2024-07-30,3868.0,1307,3,127.38,382.14,pood,Tallinn,sularaha,Anu,Paju,anu.paju@yahoo.com,+372 8407 9765,Tallinn,2020-09-30,NaN,1978.0
10125,2641,INV-202308-00218,2023-08-15,3964.0,1197,2,152.85,305.70,pood,Tartu,kaart,Kersti,Männik,kersti.mannik@yahoo.com,+372 5850 6802,Tartu,2022-04-10,bronze,1980.0


In [8]:
df = df.drop_duplicates()

In [9]:
df.shape

(11148, 19)

In [11]:
df['sale_date'] = pd.to_datetime(
    df['sale_date'],
    format='mixed',
    dayfirst=True
)

In [12]:
df['sale_date'].dtype

dtype('<M8[us]')

In [13]:
df['customer_id'].isna().sum()

np.int64(1236)

In [14]:
df = df.dropna(subset=['customer_id'])

In [15]:
df.shape

(9912, 19)

In [6]:
df.duplicated(subset=['invoice_id']).sum()

np.int64(5116)

In [7]:
df = df.drop_duplicates(subset=['invoice_id'], keep='first')

In [19]:
df.shape

(10118, 19)

In [8]:
df.isnull().sum()

sale_id                 0
invoice_id              0
sale_date               0
customer_id           988
product_id              0
quantity                0
unit_price              0
total_price             0
channel                 0
store_location       3462
payment_method          0
first_name            988
last_name             988
email                1944
phone                 988
city                  988
registration_date     988
loyalty_tier         4660
birth_year            988
dtype: int64

In [21]:
df = df.dropna(subset=['customer_id', 'sale_date', 'total_price'])

In [22]:
df['sale_date'] = pd.to_datetime(
    df['sale_date'],
    format='mixed',
    dayfirst=True
)

In [23]:
(df['total_price'] < 0).sum()

np.int64(180)

In [24]:
df = df[df['total_price'] > 0]

In [25]:
df.shape

(8950, 19)

In [26]:
df['customer_id'].nunique()

2540

In [27]:
df['sale_date'].min(), df['sale_date'].max()

(Timestamp('2023-01-01 00:00:00'), Timestamp('2026-12-03 00:00:00'))

## Roll B – Andmete puhastamine

Puhastamise käigus:
- eemaldasin 5116 duplikaati `invoice_id` järgi;
- eemaldasin 988 rida, kus puudus `customer_id`;
- teisendasin `sale_date` kuupäevaformaati;
- eemaldasin 180 negatiivse `total_price` väärtusega rida.

Pärast puhastamist:
- ridu: 8950
- veerge: 19
- unikaalseid kliente: 2540
- kuupäevavahemik: 01.01.2023–03.12.2026

Puhastatud DataFrame on valmis RFM-analüüsiks.

In [2]:
import pandas as pd

In [3]:
today = pd.to_datetime('2026-12-04')
today

Timestamp('2026-12-04 00:00:00')

In [9]:
df = df.dropna(subset=['customer_id', 'sale_date', 'total_price'])

In [10]:
df['sale_date'] = pd.to_datetime(df['sale_date'], format='mixed', dayfirst=True)

In [11]:
(df['total_price'] < 0).sum()

np.int64(180)

In [12]:
df = df[df['total_price'] > 0]

In [13]:
recency = df.groupby('customer_id')['sale_date'].max().reset_index()
recency.columns = ['customer_id', 'last_purchase']

recency['recency_days'] = (today - recency['last_purchase']).dt.days

recency.head()

,customer_id,last_purchase,recency_days
0,2001.0,2024-11-29,735
1,2004.0,2024-12-19,715
2,2005.0,2024-09-17,808
3,2006.0,2023-09-11,1180
4,2007.0,2025-01-30,673


In [14]:
frequency = df.groupby('customer_id').size().reset_index(name='frequency')

frequency.head()

,customer_id,frequency
0,2001.0,2
1,2004.0,2
2,2005.0,4
3,2006.0,1
4,2007.0,1


In [15]:
monetary = df.groupby('customer_id')['total_price'].sum().reset_index()
monetary.columns = ['customer_id', 'monetary']

monetary.head()

,customer_id,monetary
0,2001.0,203.92
1,2004.0,1198.56
2,2005.0,959.60
3,2006.0,327.06
4,2007.0,318.63


In [16]:
rfm = recency[['customer_id', 'recency_days']].merge(
    frequency, on='customer_id'
).merge(
    monetary, on='customer_id'
)

rfm.head()

,customer_id,recency_days,frequency,monetary
0,2001.0,735,2,203.92
1,2004.0,715,2,1198.56
2,2005.0,808,4,959.60
3,2006.0,1180,1,327.06
4,2007.0,673,1,318.63


In [17]:
rfm['R_score'] = pd.qcut(
    rfm['recency_days'],
    5,
    labels=[5, 4, 3, 2, 1]
)

rfm['F_score'] = pd.qcut(
    rfm['frequency'].rank(method='first'),
    5,
    labels=[1, 2, 3, 4, 5]
)

rfm['M_score'] = pd.qcut(
    rfm['monetary'],
    5,
    labels=[1, 2, 3, 4, 5]
)

rfm.head()

,customer_id,recency_days,frequency,monetary,R_score,F_score,M_score
0,2001.0,735,2,203.92,4,1,1
1,2004.0,715,2,1198.56,4,1,4
2,2005.0,808,4,959.60,3,4,3
3,2006.0,1180,1,327.06,1,1,1
4,2007.0,673,1,318.63,5,1,1


In [18]:
rfm['R_score'] = rfm['R_score'].astype(int)
rfm['F_score'] = rfm['F_score'].astype(int)
rfm['M_score'] = rfm['M_score'].astype(int)

rfm['RFM_Score'] = (
    rfm['R_score'] +
    rfm['F_score'] +
    rfm['M_score']
)

rfm.head()

,customer_id,recency_days,frequency,monetary,R_score,F_score,M_score,RFM_Score
0,2001.0,735,2,203.92,4,1,1,6
1,2004.0,715,2,1198.56,4,1,4,9
2,2005.0,808,4,959.60,3,4,3,10
3,2006.0,1180,1,327.06,1,1,1,3
4,2007.0,673,1,318.63,5,1,1,7


In [19]:
def get_segment(score):
    if score >= 13:
        return 'VIP Champions'
    elif score >= 10:
        return 'Loyal Customers'
    elif score >= 7:
        return 'Potential Loyalists'
    elif score >= 4:
        return 'At Risk'
    else:
        return 'Lost'

rfm['Segment'] = rfm['RFM_Score'].apply(get_segment)

rfm.head()

,customer_id,recency_days,frequency,monetary,R_score,F_score,M_score,RFM_Score,Segment
0,2001.0,735,2,203.92,4,1,1,6,At Risk
1,2004.0,715,2,1198.56,4,1,4,9,Potential Loyalists
2,2005.0,808,4,959.60,3,4,3,10,Loyal Customers
3,2006.0,1180,1,327.06,1,1,1,3,Lost
4,2007.0,673,1,318.63,5,1,1,7,Potential Loyalists


In [20]:
segment_counts = rfm['Segment'].value_counts().reset_index()
segment_counts.columns = ['Segment', 'Klientide arv']

segment_counts['Osakaal %'] = (
    segment_counts['Klientide arv'] / segment_counts['Klientide arv'].sum() * 100
).round(2)

segment_counts

,Segment,Klientide arv,Osakaal %
0,Potential Loyalists,743,29.25
1,Loyal Customers,691,27.20
2,At Risk,517,20.35
3,VIP Champions,461,18.15
4,Lost,128,5.04
